### PR TITLE
Implemented service to monitor the Local Worker

### PR DESCRIPTION
**Pull request by gogo4ds**
Create a Windows service to monitor the OJS Local Worker Service
Closes https://github.com/SoftUni-Internal/suls-issues/issues/1811

Send Email/SMS when Local Worker Service is down unexpectedly
Closes https://github.com/SoftUni-Internal/suls-issues/issues/4101

### CODE REVIEW

File: Open Judge System/OJS.Common/Helpers/ServicesHelper.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            var services = ServiceController.GetServices(Environment.MachineName);
+            return services.Any(s => s.ServiceName == serviceName);
+        }
+
+        private static ServiceState ConvertStatusToServiceState(ServiceControllerStatus serviceStatus)
+        {
+            switch (serviceStatus)
+            {
+                case ServiceControllerStatus.ContinuePending:
+                    return ServiceState.ContinuePending;
```
[the remark this answers is not in the document]
Comment (gogo4ds) [reply Open Judge System/OJS.Common/Helpers/ServicesHelper.cs]: I added two more states, which are not supported by the ServiceControllerStatus. Also If I simply cast the type, there is a risk or wrong state if someone changes the values of the custom type.

File: Open Judge System/Workers/OJS.Workers.Common/AppSettings/Settings.cs [hunk 2]
```diff
@@ -1,11 +1,11 @@
-﻿namespace OJS.Workers.Common
+﻿namespace OJS.Workers.Common.AppSettings
 {
     using System;
     using System.Configuration;
 
     using log4net;
 
-    public static class Settings
+    public static partial class Settings
```
Comment (jicata) [inline Open Judge System/Workers/OJS.Workers.Common/AppSettings/Settings.cs]: Why is this class `partial`? I think keeping it `static` is a better idea, given that it's literally just settings class

File: Open Judge System/Workers/OJS.Workers.LocalWorkerMonitoring/App.config [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<configuration>
+  <appSettings>
+    <add key="EmailServerHost" value="smtp.softuni.com" />
+    <add key="EmailServerPort" value="25" />
+    <add key="EmailServerUsername" value="TestUser" />
+    <add key="EmailServerPassword" value="Test" />
+    <add key="EmailSenderEmail" value="test@test.test" />
+    <add key="EmailSenderDisplayName" value="Test" />
```
Comment (jicata) [inline Open Judge System/Workers/OJS.Workers.LocalWorkerMonitoring/App.config]: Are these credentials valid for production?

Comment (gogo4ds) [reply Open Judge System/Workers/OJS.Workers.LocalWorkerMonitoring/App.config]: No, the production credentials should not be committed to the repository, but deployed manually only to the server.

File: Open Judge System/Tests/OJS.Workers.Compilers.Tests/MsBuildCompilerTests.cs [hunk 4]
```diff
@@ -5,17 +5,17 @@
...
     {
         private const string MsBuildCompilerPath = @"C:\Windows\Microsoft.NET\Framework64\v4.0.30319\MSBuild.exe";
+        private const int MsBuildCompilerProcessExitTimeOutMultiplier = 2;
 
         [Test]
         public void MsBuildCompilerShouldWorkWhenGivenValidZippedSolution()
         {
-            var compiler = new MsBuildCompiler(Settings.MsBuildCompilerProcessExitTimeOutMultiplier);
+            var compiler = new MsBuildCompiler(MsBuildCompilerProcessExitTimeOutMultiplier);
             var result = compiler.Compile(MsBuildCompilerPath, FileHelpers.SaveByteArrayToTempFile(this.GetSampleSolutionFile()), string.Empty);
```
Comment (achobanov) [inline Open Judge System/Tests/OJS.Workers.Compilers.Tests/MsBuildCompilerTests.cs]: Params on new line..

File: Open Judge System/Tests/OJS.Workers.Compilers.Tests/MsBuildCompilerTests.cs [hunk 5]
```diff
@@ -26,7 +26,7 @@ public void MsBuildCompilerShouldWorkWhenGivenValidZippedSolution()
         [Test]
         public void MsBuildCompilerShouldWorkWhenGivenValidZippedProjectInSingleFolder()
         {
-            var compiler = new MsBuildCompiler(Settings.MsBuildCompilerProcessExitTimeOutMultiplier);
+            var compiler = new MsBuildCompiler(MsBuildCompilerProcessExitTimeOutMultiplier);
             var result = compiler.Compile(MsBuildCompilerPath, FileHelpers.SaveByteArrayToTempFile(this.GetSampleSolutionWhereTheProjectIsLocatedInSingleFolder()), string.Empty);
```
Comment (achobanov) [inline Open Judge System/Tests/OJS.Workers.Compilers.Tests/MsBuildCompilerTests.cs]: Params on new line..

File: Open Judge System/Workers/OJS.Workers.LocalWorker/LocalWorkerService.cs [hunk 6]
```diff
@@ -123,5 +128,38 @@ private void CreateExecutionStrategiesWorkingDirectory()
...
+                {
+                    logger.Info($"{monitoringServiceName} is running.");
+                    return;
+                }
+
+                logger.Info($"Attempting to start the {monitoringServiceName}...");
+
+                if (serviceState.Equals(ServiceState.NotFound))
+                {
+                    const string monitoringServiceExePath =
```
Comment (jicata) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/LocalWorkerService.cs]: Can't we extract this const into a const file or something similar?

File: Open Judge System/Workers/OJS.Workers.LocalWorkerMonitoring/LocalWorkerMonitoringService.cs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+            ServicesHelper.StartService(Constants.LocalWorkerServiceName);
+
+            logger.Info($"{Constants.LocalWorkerServiceName} started successfully.");
+        }
+
+        private static void InstallAndStartLocalWorker()
+        {
+            const string localWorkerExePath =
+                @"..\..\..\OJS.Workers.LocalWorker\bin\Debug\OJS.Workers.LocalWorker.exe";
```
Comment (jicata) [inline Open Judge System/Workers/OJS.Workers.LocalWorkerMonitoring/LocalWorkerMonitoringService.cs]: Can't we extract this const into a const file or something similar?

**Comment by gogo4ds** [comment]
**Post-Deploy Action:** Add the correct email settings in the App.config of LocalWorkerMonitoringService

**Pull request merged**